source/robot_model/src/Model.cpp

```diff
@@ -5,277 +5,381 @@
-  // the velocity vector contains position of the intermediate frame and full pose of the end-effector
-  Eigen::VectorXd delta_r(3 * cartesian_states.size() + 3);
-  Eigen::MatrixXd jacobian(3 * cartesian_states.size() + 3, nb_joints);
-
-  for (unsigned int i = 0; i < cartesian_states.size() - 1; ++i) {
-    CartesianState state = cartesian_states[i];
-    // extract only the position for intermediate points
-    delta_r.segment<3>(3 * i) = state.get_linear_velocity();
-    jacobian.block(3 * i, 0, 3 * i + 3, nb_joints) =
-        this->compute_jacobian(joint_state, state.get_name()).data().block(0, 0, 3, nb_joints);
+
+  StateRepresentation::CartesianPose Model::forward_geometry(const StateRepresentation::JointState &joint_state,
+                                                             std::string frame_name)
+  {
+    if (frame_name.empty())
+    {
+      // get last frame if none specified
+      frame_name = this->robot_model_.frames.back().name;
+    }
+    return this->forward_geometry(joint_state, std::vector<std::string>{frame_name}).front();
   }
-  // extract the orientation for the end-effector
-  CartesianState state = cartesian_states.back();
-  delta_r.segment<3>(3 * (cartesian_states.size() - 1)) = state.get_linear_velocity();
-  delta_r.tail(3) = state.get_angular_velocity();
-  jacobian.bottomRows(6) = this->compute_jacobian(joint_state, state.get_name()).data();
-  // compute the Jacobian
-  Eigen::MatrixXd hessian_matrix = jacobian.transpose() * jacobian;
-
-  // set the hessian sparse matrix
-  std::vector<Eigen::Triplet<double>> coefficients;
-  for (unsigned int i = 0; i < nb_joints; ++i) {
-    for (unsigned int j = 0; j < nb_joints; ++j) {
-      coefficients.emplace_back(Eigen::Triplet<double>(i, j, hessian_matrix(i, j)));
+
+  StateRepresentation::JointPositions Model::inverse_geometry(StateRepresentation::JointState actual_joint_state,
+                                                              const StateRepresentation::CartesianState &desired_cartesian_state)
+  {
+    // This method uses the angular velocity evaluated by the 'Infeasibility-free inverse kinematics method' which takes into account the
+    // joint limits. The angular velocity is then used for indicating the direction for an iterative algorithm.
+    // To converge at the desired end-effector position a similar approach to the Newton–Raphson method is used, the only difference is that
+    // here the pseudo-inverse of the jacobian is not used and instead is used the angualr velocity evaluated by the
+    // 'Infeasibility-free inverse kinematics method'.
+
+    const int JOINT_ID = this->robot_model_.njoints - 2;
+
+    // TODO if the actual_joint_state is not provided initialize the position of the robot in a radom conformation
+    // TODO take as input the parameters eps and IT_MAX
+    // TODO return 'success'
+
+    Eigen::VectorXd q = actual_joint_state.get_positions(); //pinocchio::neutral(model);
+    const double eps = 1e-4;
```

Comment:
Are these magic numbers or how did you set them? I'm curious to know how many iterations it actually takes. (This is not a criticism, I'm just curious)

Reply:
Magic numbers, they have to be tuned. I will probably change them when I do a the google test